### DISCUSSION TITLE
Upgradeable `RwLock` read locks?

### DISCUSSION THREAD

**Discussion by Jules-Bertholet** [tokio-rs/tokio#5718] [Ideas]
`parking_lot`'s `RwLock` has upgradeable read locks, that can be upgraded to a write lock without letting other writers through in the meantime. Has any thought been given to supporting this in tokio's `RwLock`?

**Comment by Darksonn** [comment]
I don't see how that could work. You can have multiple read guards, so trying to upgrade two coexisting guards must necessarily fail.

**Reply by Jules-Bertholet** [reply]
What `parking_lot` does is to have a separate type for upgradeable and non-upgradeable read guards, with only one upgradeable guard allowed to exist at any one time. ([doc link](https://docs.rs/lock_api/0.4.9/lock_api/struct.RwLock.html#method.upgradable_read))

**Reply by Darksonn** [reply]
I see. I don't think we're going to add an entirely new type just for that.

**Reply by Jules-Bertholet** [reply]
To give an idea of my use-case:

I have an object, where most of the time, only one task is going to be managing it (reads + writes) for a long stretch of time. But I have some separate monitor tasks that will want to briefly read from the object at regular intervals. My thinking was to give the "owning" task an upgradeable read lock, which it can upgrade and downgrade at its pleasure, while the monitor tasks take standard read locks when they need to.

**Reply by Jules-Bertholet** [reply]
Oh, it seems that the [`async-lock` crate](https://docs.rs/async-lock/2.7.0/async_lock/struct.RwLock.html#method.upgradable_read) offers this, I think I will just use it for now.
